src/core/net/dnssd_server.cpp

```diff
@@ -1287,6 +1386,574 @@ void Server::ResetUpstreamQueryTransaction(UpstreamQueryTransaction &aTxn, Error
+    instanceInfo.mServiceType     = serviceType;
+    instanceInfo.mServiceInstance = instanceLabel;
+    instanceInfo.mInfraIfIndex    = Get<BorderRouter::InfraIf>().GetIfIndex();
+
+    switch (aCommand)
+    {
+    case kStart:
+        Get<Dnssd>().StartServiceResolver(instanceInfo);
+        break;
+
+    case kStop:
+        Get<Dnssd>().StopServiceResolver(instanceInfo);
+        break;
+    }
+}
+
+void Server::DiscoveryProxy::UpdateIp6AddressResolver(Command aCommand, const DnsName &aHostName)
+{
+    // Start or stop an IPv6 address resolver for a given host name
+    // on infra-if network.
+
+    DnsName  hostLabels;
+    uint32_t infraIfIndex;
+
+    IgnoreError(StripDomainName(aHostName, hostLabels));
+    infraIfIndex = Get<BorderRouter::InfraIf>().GetIfIndex();
+
+    switch (aCommand)
+    {
+    case kStart:
+        Get<Dnssd>().StartIp6AddressResolver(hostLabels, infraIfIndex);
+        break;
+
+    case kStop:
+        Get<Dnssd>().StopIp6AddressResolver(hostLabels, infraIfIndex);
+        break;
+    }
+}
+
+bool Server::DiscoveryProxy::QueryMatches(const ProxyQuery     &aQuery,
+                                          const ProxyQueryInfo &aInfo,
+                                          ProxyAction           aAction,
+                                          const DnsName        &aName) const
+{
+    // Check whether `aQuery` is performing `aAction` and
+    // its name matches `aName`.
+
+    bool matches = false;
+
+    VerifyOrExit(aInfo.mAction == aAction);
+
+    switch (aAction)
+    {
+    case kBrowsing:
+        VerifyOrExit(QueryNameMatches(aQuery, aName));
+        break;
+    case kResolvingService:
+        VerifyOrExit(QueryInstanceNameMatches(aQuery, aInfo, aName));
+        break;
+    case kResolvingIp6Address:
+        VerifyOrExit(QueryHostNameMatches(aQuery, aInfo, aName));
+        break;
+    case kNoAction:
+        ExitNow();
+    }
+
+    matches = true;
+
+exit:
+    return matches;
+}
+
+bool Server::DiscoveryProxy::HasActive(ProxyAction aAction, const DnsName &aName) const
+{
+    // Determine whether or not we have an active browser/resolver
+    // corresponding to `aAction` for `aName`.
+
+    bool has = false;
+
+    for (const ProxyQuery &query : Get<Server>().mProxyQueries)
+    {
+        ProxyQueryInfo info;
+
+        info.ReadFrom(query);
+
+        if (QueryMatches(query, info, aAction, aName))
+        {
+            has = true;
+            break;
+        }
+    }
+
+    return has;
+}
+
+void Server::DiscoveryProxy::HandleServiceBrowseResult(Dnssd::Event                  aEvent,
+                                                       const Dnssd::ServiceInstance &aServiceInstance)
+{
+    DnsName serviceName;
+
+    VerifyOrExit(mIsRunning);
+    VerifyOrExit(aEvent == Dnssd::kEventAdded);
+    VerifyOrExit(aServiceInstance.mInfraIfIndex == Get<BorderRouter::InfraIf>().GetIfIndex());
+
+    ConstructFullName(aServiceInstance.mServiceType, serviceName);
```

Comment:
For subtype browsing, should this function also append the subtype string + ```._sub.```? For a generic browse, like ```_test._tcp``` a matching query from DiscvoverProxy is found in ```Server::DiscoveryProxy::HandleResult()```, but for a browse like ```_printer._test._tcp```, a matching query seems not to be discovered in this current implementation.

Reply:
I think we expect the `mServiceType` in `ServiceInstance` to include it already (if it is for sub-type service):
```c++
typedef struct otPlatDnssdServiceInstance
{
    const char *mServiceType;     ///< Service type or sub-type (e.g., "_mt._udp", "_s1._sub._mt._udp")
    const char *mServiceInstance; ///< Service instance label.
    uint32_t    mTtl;             ///< TTL in seconds.
    uint32_t    mInfraIfIndex;    ///< The infrastructure network interface index.
} otPlatDnssdServiceInstance;
```

BTW, we may need to change the platform APIs defined in this PR to aligen them more closely with the definitions in https://github.com/openthread/openthread/pull/9797

Reply:
Got it! Platform code has to create ```mServiceType``` with corresponding labels before notifying DiscoverProxy. Thanks!